Automattic/jetpack
# WordAds: Add consent support for California Consumer Privacy Act (CCPA)

#### Changes proposed in this Pull Request:

Adds support for California site visitors to opt-out of targeted advertising as specified in the California Consumer Privacy Act (applies to WordAd ads).

* Adds settings to Jetpack->Settings->Traffic to enable/disable the feature
* Adds `[ccpa-do-not-sell-link]` shortcode and widget to display required "Do Not Sell My Personal Information" link
* Adds Modal Notice popup with CCPA notice, and option for site visitors to opt-out of targeted ads
* Handles setting appropriate first party cookies to maintain users opt-out status
* Geo-locates site visitors so "Do Not Sell" link is only displayed to California IP addresses

When the setting is enabled, and the widget/shortcode added to the site, two additional cookies are set for site visitors:

1. `ccpa_applies`: Stores the result of the geolocation test for 24 hours. True/false if the visitor's IP address resolves to California.

2. `usprivacy`: Stores the visitor's consent selection in a format specified by the IAB. Saved for one year if opting in to targeted advertising, and five years if opting out of targeted advertising. This cookie is read and consent passed to advertisers as part of the ad bidding process.



Settings Panel:

![Screen Shot on 2020-04-02 at 14-30-41](https://user-images.githubusercontent.com/1050616/78301852-92abea00-74ee-11ea-96a2-59ca7b3f56fa.png)

Modal Notice:

![Screen Shot on 2020-04-02 at 14-31-39](https://user-images.githubusercontent.com/1050616/78301984-d1da3b00-74ee-11ea-8baa-5dcfb1900d59.png)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?

This is a new feature.
Internal reference: pMyNb-4Rh-p2

#### Testing instructions:

Note: These tests must be done from a California IP address. You can either use a VPN, or set a site cookie `ccpa_applies=true`.

* Create a new Jetpack site with a Premium plan or higher, and enrolled in WordAds
* Enable the CCPA option
* Place the Do Not Sell widget or shortcode on the homepage (likely in the footer area)
* Verify the "Do Not Sell My Personal Information" link appears
* Click the link, and verify the Modal Notice appears
* Verify the `usprivacy` cookie is set to `1YNN`
* Opt-out from the Modal Notice
* Verify the `usprivacy` cookie is set to `1YYN`

#### Proposed changelog entry for your changes:

* Jetpack Ads: Site owners can enable a "Do Not Sell My Personal Information" link as specified in the California Consumer Privacy Act (CCPA) for California site visitors to opt-out of targeted advertising.
